### PR TITLE
Don't add dependencies recursively while linking

### PR DESCRIPTION
**Pull request by nirbheek**
We were doing this on the basis of an old comment, but there was no test for it and I couldn't reproduce the issue with clang on Linux at all.

Let's add a test and see if it breaks anywhere if we stop doing this.

Halves the size of gstreamer's build.ninja from 19M to 8.5M

Closes https://github.com/mesonbuild/meson/issues/1057

### CODE REVIEW

**Comment by tp-m** [comment]
It appears that with this patch in the gtk4 meson build, the static gdk backend libs are no longer linked into the static lib libgdk.a now. They were passed to `static_library(..., link_with: ...)`

**Comment by jpakkane** [comment]
We really want this in as soon as possible. This may be the last chance to change this before we have too many users.

**Comment by jpakkane** [comment]
Apparently we don't have a test for exe -> static -> static, only one for exe -> static-> shared. We should add one.

**Comment by nirbheek** [comment]
Yeah, it's on my TODO list for this PR.

**Comment by keszybz** [comment]
FWIW, I'm building systemd with meson with this patch, and it seems to work nicely.

**Comment by nirbheek** [comment]
I should have some time to work on this tomorrow. If we have enough tests for various corner-cases, we could merge this for 0.40, and fix any breakage in a point release?

**Comment by jpakkane** [comment]
This message is posted to every outstanding merge request.

We have transitioned from Github wiki to our new documentation system that generates all documentation directly from the Git repo. This means that from now on every merge request must come with corresponding documentation changes.

If your MR requires documentation changes, do the necessary changes to files in `docs/markdown`.

If your MR adds a new feature, add a description to the release notes in `docs/markdown/Release-notes-for-0.41.0.md`.

If this commit is just a bugfix with no functional changes, you don't need to do anything.

Thank you

**Comment by jpakkane** [comment]
What's the status of this? Is any help needed?

**Comment by nirbheek** [comment]
Dropped off my list accidentally, will look into this today!

File: mesonbuild/compilers.py [hunk 1]
```diff
@@ -2402,6 +2433,15 @@ def get_compiler_is_linuxlike(compiler):
         return True
     return False
 
+def get_compiler_uses_gnuld(c):
+    # FIXME: Perhaps we should detect the linker in the environment?
+    # FIXME: Assumes that *BSD use GNU ld, but they might start using lld soon
+    if (getattr(c, 'gcc_type', None) in (GCC_STANDARD, GCC_MINGW, GCC_CYGWIN)) or \
+       (getattr(c, 'clang_type', None) in (CLANG_STANDARD, CLANG_WIN) or \
+       (getattr(c, 'icc_type', None) in (ICC_STANDARD, ICC_WIN)):
```
Comment (ignatenkobrain) [inline mesonbuild/compilers.py]: __[flake8]__
* [E999] SyntaxError: invalid syntax

**Comment by nirbheek** [comment]
"Today" came 18 days late, but I have something that works for most cases now. Some people will get build failures because they under-specified the libraries that they need, but gstreamer itself didn't need any such fixes and that makes me think that people are generally disciplined about this.

Anyway, the fix in such cases is easy and is backwards-compatible so I think we shouldn't worry about this too much.

File: mesonbuild/backend/ninjabackend.py [hunk 2]
```diff
@@ -24,16 +29,13 @@
 from ..mesonlib import get_meson_script, get_compiler_for_source
 from .backends import CleanTrees, InstallData
 from ..build import InvalidArguments
-import os, sys, pickle, re
-import subprocess, shutil
-from collections import OrderedDict
 
 if mesonlib.is_windows():
-    quote_char = '"'
+    quote_func = lambda s: '"{}"'.format(s)
```
Comment (ignatenkobrain) [inline mesonbuild/backend/ninjabackend.py]: __[flake8]__
* [E731] do not assign a lambda expression, use a def

File: mesonbuild/backend/ninjabackend.py [hunk 3]
```diff
@@ -105,18 +107,17 @@ def write(self, outfile):
...
-            q_templ = quote_char + "%s" + quote_char
             noq_templ = "%s"
             newelems = []
             for i in elems:
                 if not should_quote or i == '&&': # Hackety hack hack
-                    templ = noq_templ
+                    quoter = ninja_quote
                 else:
-                    templ = q_templ
+                    quoter = lambda x: ninja_quote(quote_func(x))
```
Comment (ignatenkobrain) [inline mesonbuild/backend/ninjabackend.py]: __[flake8]__
* [E731] do not assign a lambda expression, use a def

**Comment by nirbheek** [comment]
Also, here's some stats:

[glib](https://git.gnome.org/browse/glib/log/?h=wip/meson)'s `build.ninja` went from 614KB to 531KB
[gst-build](https://cgit.freedesktop.org/gstreamer/gst-build/)'s `build.ninja` went from 20MB to 6.0MB

All this with no significant changes in configure time (gst-build's configure time went down by 10% for me, but likely because of the smaller `build.ninja` file).

These numbers can be reduced a bunch more by using Ninja rules for each target so that build statements don't have to repeat the cflags over and over for each source file in the target. We could reduce this even more by putting compiler-specific cflags in the compiler rules instead of in each target's build statements.

Interestingly, the "use shlex.quote for quoting" commit in this PR also reduces the `build.ninja` size considerably: 100KB for gst-build.

**Comment by jpakkane** [comment]
executable('circular', 'main.c', link_with : [st1, st2, st3])

I'm surprised that this works on platforms without groupings, mainly MSVC. IIRC it does something saner than LD by looking at past libraries for symbols. So the question is:

 - is MSVC guaranteed to do the right thing
 - do we ever need to support a platform that gets this wrong _and_ does not have static library grouping?

If the latter is true (dunno, maybe Solaris or somesuch works like that) then having the dependencies a a set is problematic because you would need have something like `[st1, st2, st3, st1, st2]` go through to the link line unaltered.

We could also say that this is too niche and we don't care or will fix it later if the issue really arises.

**Comment by nirbheek** [comment]
I wrote the `circular` test specifically because I had read that only GNU `ld` is dumb while resolving symbols and I wanted to add a test for it. macOS and MSVC are saner, fwict.

Perhaps Sun Studio's `ld` has the same issue as GNU `ld`, but we've never tested anything with that, so we should deal with that when someone cares about it. I will note that OpenIndiana moved from Sun Studio to GCC last year, so we may not have to care about it ever.

> because you would need have something like `[st1, st2, st3, st1, st2]` go through to the link line unaltered.

Specifying a library multiple times will now be ignored because we de-dup library names in the `CompilerArgs` class while constructing compiler arguments. This can be turned off for a broken linker in the future if needed.

**Comment by nirbheek** [comment]
Let's list the platforms that we actively support:

1. Linux (GNU ld)
2. Windows (MSVC link, GNU ld, Clang lld)
3. macOS and iOS (Apple ld, Clang lld)
4. Android (GNU ld)
5. *BSD (GNU ld, Clang lld(?))

Others:

1. OpenIndiana (GNU ld)
2. Haiku OS (GNU ld)

Is there anything else we want to support? I think most edge-platforms actually rely on GCC, so we shouldn't have much to worry about.

**Comment by QuLogic** [comment]
gold?

**Comment by nirbheek** [comment]
I tested gold, it behaves the same way as bfd at least on Fedora.

**Comment by jpakkane** [comment]
🏁 🎉 🎆 🥂

**Pull request merged**